### PR TITLE
Load on frontend

### PR DESCRIPTION
**Pull request by pdewouters**
Would it be possible to run only on the backend's *backupwordpress* pages?

- not on the frontend
- not on other admin pages

```php
define( 'WHATS_RUNNING_HIGHLIGHT', 'plugins/backupwordpress' );
```
Tool: https://wordpress.org/plugins/whats-running/screenshots/

Thank you!

### CODE REVIEW

**Comment by pdewouters** [comment]
still need to check other admin pages

**Comment by szepeviktor** [comment]
You could also go *deep* and analyze every one of these entry points with What"s running:
https://github.com/szepeviktor/WPHW/blob/master/wp-entry-points.md#wordpress-entry-points-wp-loadphp-occurrences

**Comment by willmot** [comment]
Do cron requests run under WP-admin?

**Comment by szepeviktor** [comment]
No, those are `defined('DOING_CRON') && DOING_CRON`

**Comment by pdewouters** [comment]
hmm ,we got a bit of a dependency nightmare here :D

**Comment by szepeviktor** [comment]
Maybe http://php.net/manual/en/function.spl-autoload-register.php could solve it.

**Comment by pdewouters** [comment]
@szepeviktor yes we have an open issue for adding an autoloader actually.

File: backupwordpress.php [hunk 1]
```diff
@@ -33,6 +33,10 @@
 
 namespace HM\BackUpWordPress;
 
+if ( ! is_admin() || ( defined( 'DOING_CRON' ) && DOING_CRON ) || ( defined( 'DOING_AJAX' ) && DOING_AJAX ) ) {
```
Comment (willmot) [inline backupwordpress.php]: Logic is wrong here, we should `return` is `! DOING_CRON!`, I don't think we'd need to check for `DOING_AJAX` as `admin-ajax.php` already runs in admin context.

**Comment by pdewouters** [comment]
@willmot actually there is no need to do this and can actually cause more issues than benefits. I realized this after chatting to @joehoyle 
We can still review / merge this as I ended up fixing some logic and formatting.

**Comment by szepeviktor** [comment]
I got very sad when I saw your plugin on the frontend.

**Comment by szepeviktor** [comment]
I distinguish *popular* plugins from *enterprise* plugins by speed and security and maybe unnecessary features.
I personally prefer fast, secure any low-on-features plugins.

**Comment by szepeviktor** [comment]
@pdewouters Have any of your tests failed after `is_admin`-ing your plugin?

**Comment by pdewouters** [comment]
szepeviktor feel free to submit a PR with some stats if you manage to improve the performance.

**Comment by szepeviktor** [comment]
What does [WIP] stand for?

**Comment by szepeviktor** [comment]
This is the proposed format:
```php
array(4) {
  [0]=>
  array(11) {
    ["summary"]=>
    string(18) ""1" post published"
    ["author_meta.user_login"]=>
    string(6) "viktor"
    ["author_role"]=>
    string(13) "administrator"
    ["created"]=>
    string(28) "2015-02-25T18:07:20.709+0000"
    ["author"]=>
    int(1)
    ["ip"]=>
    string(14) "192.168.12.146"
    ["meta.links.self"]=>
    string(125) "https://api.wp-stream.com/sites/site-98b7bac0-bd18-11e4-952d-1b5d1e45d3ae/records/record-213c4eb0-bd19-11e4-952d-1b5d1e45d3ae"
    ["meta.links.collection"]=>
    string(81) "https://api.wp-stream.com/sites/site-98b7bac0-bd18-11e4-952d-1b5d1e45d3ae/records"
    ["meta.score"]=>
    NULL
    ["meta.sort"]=>
    int(1424887640709)
    ["ID"]=>
    string(43) "record-213c4eb0-bd19-11e4-952d-1b5d1e45d3ae"
  }
  [1]=>
  array(11) {
    ["summary"]=>
    string(16) ""1" post drafted"
    ["author_meta.user_login"]=>
    string(6) "viktor"
    ["author_role"]=>
    string(13) "administrator"
    ["created"]=>
    string(28) "2015-02-25T18:07:18.045+0000"
    ["author"]=>
    int(1)
    ["ip"]=>
    string(14) "192.168.12.146"
    ["meta.links.self"]=>
    string(125) "https://api.wp-stream.com/sites/site-98b7bac0-bd18-11e4-952d-1b5d1e45d3ae/records/record-1fabeab0-bd19-11e4-952d-1b5d1e45d3ae"
    ["meta.links.collection"]=>
    string(81) "https://api.wp-stream.com/sites/site-98b7bac0-bd18-11e4-952d-1b5d1e45d3ae/records"
```

**Comment by szepeviktor** [comment]
This is the documentation. I suggest to keep it outside the code.
```php
// Filtering fields is possible by configuring it in wp-cli.yml
// stream query:
//    fields: created,ip,author,author_meta.user_login,author_role,summary
```

**Comment by pdewouters** [comment]
WIP = Work In Progress

**Comment by szepeviktor** [comment]
> feel free to submit a PR with some stats if you manage to improve the performance.

Do you mean you are not aware of PHP parsing and execution times?

I suspect that speed optimized websites use server-side backup and do not have maintenance plugins turned on. I could be wrong here.
Almost all of my customers have VPS-s.

**Comment by joehoyle** [comment]
@szepeviktor do you have any data to show BackUpWordPress is taking significant time to parse? PHP 5.4+ (I think) has a built it opcode cache, and php-apc is in pretty common usage before that. Also, disk IO is pretty low due to the opcode cache again, and system level file caching.

In my experience, reducing "files loaded" is poorly spent optimisation time. In real world cases, that's not what is causing page generation to be slow. I'm all ears of course if we have specific data on how long including a few PHP files. 

The downside of trying to do this (IMO) premature optimisation with only loading the plugin on your specific part of the admin, or whatever is that other parts of the system can't take advantage of it. For example, imagine someone wants to add a BackUpWordPress action in XMLRPC so they can trigger backups from external events, or use the WP-API to build integration with BackUpWordPress, or someone else makes an extension to do xyz. But limiting where you load the plugin (PHP load of course, no admin css / js should be showing anywhere unnecessary)

**Comment by pdewouters** [comment]
Tests are failing because now the version check is actually working and I set the min WP version to `3`.9.3`  and in the travis config we are testing against `3.9`

**Comment by szepeviktor** [comment]
OK. I am not willing to convince you. Maybe I should tell you *this is my personal preference*. I work with small VPS-s, low on resources, and I am trying to minimize CPU load.

**Comment by szepeviktor** [comment]
Without page cache front page times look like this:
```
  1.1 ++----------------+---------------------+---------------------+---------------------+---------------------+--++
      |                 +                     +                     +                     +                     +   |
      |                 :                     :                     :                     :                     :   |
      |                 :                     :                     :                     :                     :   |
      |                 :       A             :                     :                     :                     :   |
    1 ++........................*..................................................................................++
      |                 :       *             :                     :                     :                     :   |
      |                 :      * *            :                     :                     :                     :   |
      |                 :      * *            :                     :                     :                     :   |
      |                 :      * *            :                     :                     :                     :   |
  0.9 ++.......................*.*.................................................................................++
      |                 :      * *            :                     :                     :                     :   |
      |                 :      *  *           :                     :                     :                     :   |
      |                 :     *   *           :                     :                     :                     :   |
      |                 :     *   *      A    :                     :                     :                     :   |
  0.8 ++......................*...*......*.........................................................................++
      |                 :     *   *     * *   :                     :                     :                     :   |
      |                 :     *    *    * *   :                     :                     :                     :   |
      |                 :    *     *    *  *  :                     :                     :                     :   |
      |                 :    *     *    *  *  :                     :                     :                     :   |
  0.7 ++.....................*.....*...*....*......................................................................++
      |                 :    *     *   *    * :                     :                     :                     :   |
      |                 :    *      *  *     *:                   **A                     :                     :   |
      |                 :    *      *  *     *:                A**  *                     :                     :   |
      |                 :   *       * *       *                *    :*                    :                     :   |
  0.6 ++....................*.......*.*.......A***A...........*......*.............................................++
      |                 :   A       * *       :    *          *     : *                   :                     :   |
      |                 :  *         **       :    *         *      : *          A        :                     :   |
      |                 : *          *        :     *        *      :  *       ** *       :            A        :   |
      |  *A****         : *          A        :      *      *       :  *     **    *      :           * *       :   |
  0.5 +**......A***A****.*...........................*......*...........*.**A.......*...............**...*.........*A
      A                 A                     :       A*****        :   A*           A****A***     *      *     :** |
      |                 :                     :            A        :                     :   A***A        A****A   |
      |                 :                     :                     :                     :                     :   |
      |                 +                     +                     +                     +                     +   |
  0.4 ++----------------+---------------------+---------------------+---------------------+---------------------+--++
                       540                   545                   550                   555                   560
```

So we won't see 10 or 20 msec increase.

**Comment by szepeviktor** [comment]
I appreciate you efforts.

File: backupwordpress.php [hunk 2]
```diff
@@ -76,7 +76,7 @@ public static function get_instance() {
 	 */
 	public function plugins_loaded() {
 
-		if ( false !== $this->maybe_self_deactivate() ) {
```
Comment (willmot) [inline backupwordpress.php]: Is the logic switch here intentional?

Comment (pdewouters) [inline backupwordpress.php]: Yes - maybe the naming is confusing?
If we need to self deactivate it returns true.

Comment (willmot) [inline backupwordpress.php]: No that makes sense, just wanted to double check.
